design/host-discovery.md

```diff
@@ -0,0 +1,292 @@
+   the host
+6. Automatically provision the host as a node in the cluster
+
+#### Adding a host and not automatically provisioning
+
+*As a cluster admin, I want to add a node to a cluster without
+pre-configuring the BMC and NOT have the cluster automatically
+provision the host as a node.*
+
+This use case is different from the previous case because the
+`MachineSet` would not scale automatically and so the host would not
+be provisioned. This gives cluster admins an opportunity to review the
+host before provisioning, to ensure that it should be a member of the
+cluster and is not misconfigured (PXE booting from the wrong
+environment) or a malicious actor.
+
+#### Replacing a faulty node
+
+*As a data center staff member, I want to replace a faulty node in a
+cluster quickly.*
+
+The idea here is to be able to take a “broken” host out of service and
+replace it with a standby host without having access to the kubernetes
+API or UI at all. If we assume the broken host is broken enough that
+it is completely offline, it should be possible to remove it from the
+rack and replace it with another physical host, which is then
+provisioned automatically.
+
+## Design Details
+
+When Ironic is configured to support its existing discovery feature,
+unknown hosts that boot the Ironic agent will have their hardware
+details registered with `ironic-inspector` and have a node created in
+the Ironic database. Typically this happens via the PXE boot
+configuration, but in the future it could happen in other ways such as
+having a bootable ISO image with the agent on it. How the agent ends
+up on the host should not matter in the rest of the workflow.
+
+The data collected for an unknown host includes the same inspection
+data available for any other host, but does not include the BMC
+credentials. We could use the data to partially create a
+`BareMetalHost` resource for the host by polling the Ironic API to
+compare the list of nodes it knows about with the list of host
+resources. This polling is independent of the normal controller
+reconciliation loop, and should run in a single separate goroutine in
+the `baremetal-operator` process, started after the process wins the
+leader election. A command line flag will control the frequency of
+polling with a default of every `60` seconds. If the frequency is set
+to `0` then polling for discovered hosts will be disabled.
+
+The discovery loop will create a `BareMetalHost` with its name set to
+the Ironic UUID for the discovered node. This will ensure a unique
+name, and will allow the `IronicProvisioner` in the
+`baremetal-operator` to match the host resource to the Ironic node
+automatically. Resource names are immutable, so this may make it
+difficult for a user to keep track of hosts, but other discovered
+identifying information such as the MAC address, IP address(es), and
+serial numbers will also be available. Users can also place data in
```

Comment:
A thing I've learned from TripleO is that people care really really deeply about controlling names. The hacks we had to do to make that possible were onerous, manual, and made everything more brittle than it needed to be. But while people might be mildly irritated if everything occasionally falls over in a heap, they will go somewhere else in a heartbeat if you make them accept automatically generated names.

Reply:
Do you have any suggestions for better ways to pick identifying information?

Reply:
Nevermind, the templating idea you propose is pretty interesting, let me explore that.